### DISCUSSION TITLE
Properties of cobra commands to expose

### DISCUSSION THREAD

**Discussion by ankithans** [aerogear/charmil#40] [Ideas]
Ignored Run, Usage and other functions as exposing functions doesn't make sense for wrapper approach

### Already Exposing
- `Use string`
- `Short string`
- `Example string`
- `Args PositionalArgs`

### Can be Exposed
	
- `Aliases []string` (Aliases is an array of aliases that can be used instead of the first word in Use)
- `SuggestFor []string` (similar to aliases but only suggests)
- `Long string`
- `ValidArgs []string` (list of all valid non-flag arguments that are accepted in bash completions)
- `ArgAliases []string` (List of aliases for ValidArgs)
- `BashCompletionFunction string` (custom functions used by the bash autocompletion generator)
- `Deprecated string` (if this command is deprecated and should print this string when used)
- `Annotations map[string]string` (key/value pairs that can be used by applications to identify or group commands)
- `Version string`
- `args []string` (args is actual args parsed from flags)

### Left over
```go
	// flagErrorBuf contains all error messages from pflag.
	flagErrorBuf *bytes.Buffer
	// flags is full set of flags.
	flags *flag.FlagSet
	// pflags contains persistent flags.
	pflags *flag.FlagSet
	// lflags contains local flags.
	lflags *flag.FlagSet
	// iflags contains inherited flags.
	iflags *flag.FlagSet
	// parentsPflags is all persistent flags of cmd's parents.
	parentsPflags *flag.FlagSet

	// usageTemplate is usage template defined by user.
	usageTemplate string
	// helpTemplate is help template defined by user.
	helpTemplate string
	// helpCommand is command with usage 'help'. If it's not defined by user,
	// cobra uses default help command.
	helpCommand *Command
	// versionTemplate is the version template defined by user.
	versionTemplate string

	// inReader is a reader defined by the user that replaces stdin
	inReader io.Reader
	// outWriter is a writer defined by the user that replaces stdout
	outWriter io.Writer
	// errWriter is a writer defined by the user that replaces stderr
	errWriter io.Writer

	//FParseErrWhitelist flag parse errors to be ignored
	FParseErrWhitelist FParseErrWhitelist

	// commandsAreSorted defines, if command slice are sorted or not.
	commandsAreSorted bool
	// commandCalledAs is the name or alias value used to call this command.
	commandCalledAs struct {
		name   string
		called bool
	}

	// commands is the list of commands supported by this program.
	commands []*Command
	// parent is a parent command for this command.
	parent *Command
	// Max lengths of commands' string lengths for use in padding.
	commandsMaxUseLen         int
	commandsMaxCommandPathLen int
	commandsMaxNameLen        int

	// TraverseChildren parses flags on all parents before executing child command.
	TraverseChildren bool

	// Hidden defines, if this command is hidden and should NOT show up in the list of available commands.
	Hidden bool

	// SilenceErrors is an option to quiet errors down stream.
	SilenceErrors bool

	// SilenceUsage is an option to silence usage when an error occurs.
	SilenceUsage bool

	// DisableFlagParsing disables the flag parsing.
	// If this is true all flags will be passed to the command as arguments.
	DisableFlagParsing bool

	// DisableAutoGenTag defines, if gen tag ("Auto generated by spf13/cobra...")
	// will be printed by generating docs for this command.
	DisableAutoGenTag bool

	// DisableFlagsInUseLine will disable the addition of [flags] to the usage
	// line of a command when printing help or generating docs
	DisableFlagsInUseLine bool

	// DisableSuggestions disables the suggestions based on Levenshtein distance
	// that go along with 'unknown command' messages.
	DisableSuggestions bool

	// SuggestionsMinimumDistance defines minimum levenshtein distance to display suggestions.
	// Must be > 0.
	SuggestionsMinimumDistance int
```

**Comment by wtrocki** [comment]
> functions doesn't make sense for wrapper approach

So you telling me that wrapper approach will not allow validation.  Eg: https://github.com/redhat-developer/app-services-cli/blob/main/pkg/cmd/kafka/topic/describe/describe.go#L56

**Comment by wtrocki** [comment]
Sounds good. Although not needed to be implemented right now for POC. we needed to see list of commands

**Comment by ankithans** [comment]
> So you telling me that wrapper approach will not allow validation

like I was saying we can implement validation while creating cobra command, but no need to get that validation function from the user. We can create that func while creating command based on other properties that we have exposed.

> for POC we needed to see list of commands

means a full plugin/CLI working? for POC

**Comment by wtrocki** [comment]
I meant that while this is great research lets keep poc simple. No need to add all of those extra arguments listed in this issue right now.
We are simply building backlog
